core/src/main/java/oracle/weblogic/deploy/util/WLSDeployArchive.java

```diff
@@ -571,20 +599,27 @@ public String getFileHash(String path) throws WLSDeployArchiveIOException {
      * method assumes that the new one also needs to be added so it changes the name to prevent conflicts by adding
      * a numeric value onto the file's basename (e.g., myapp(1).ear, myapp(2).ear).
      *
-     * @param appPath - File representing the actual path of the archive or directory in the file system
+     * @param appPath - file name representing the actual path of the archive or directory in the local or remote
+     *                file system
      * @return the relative path where the application will be unpacked by the unpackApplications() method
      * @throws WLSDeployArchiveIOException if an IOException occurred while reading or writing changes
      * @throws IllegalArgumentException    if the file or directory passed in does not exist
      */
-    public String addApplication(File appPath) throws WLSDeployArchiveIOException {
+    public String addApplication(String appPath) throws WLSDeployArchiveIOException {
         final String METHOD = "addApplication";
 
         LOGGER.entering(CLASS, METHOD, appPath);
-        String newName = appPath.getPath();
-        if (hasArchive) {
-            validateExistingFile(appPath, "appPath", getArchiveFileName(), METHOD, true);
 
-            newName = addItemToZip(ARCHIVE_APPS_TARGET_DIR, appPath);
+        File filePath = new File(appPath);
+        String newName = appPath;
+        if (isRemote) {
+            newName = addItemToZip(ARCHIVE_APPS_TARGET_DIR, filePath);
+            addToRemoteMap(appPath, newName, ArchiveEntryType.APPLICATIONS.name());
+        }
+        else if (hasArchive) {
+            validateExistingFile(filePath, "appPath", getArchiveFileName(), METHOD, true);
+
+            newName = addItemToZip(ARCHIVE_APPS_TARGET_DIR, filePath);
```

Comment:
As discussed, this feels strange.  We are calling `addXXX()` methods on a class named `WLSDeployArchiveFile` even when we are not really adding anything to the archive file.  We are also calling `addItemToZip`, when there is no zip file.  To me, this makes the code harder to understand.  It seems like the abstraction level is wrong.  I would think that the `WLSDeployArchiveFile` shouldn't need to understand how to behave when we are not using an archive file.  Also, this abstraction will likely need to be refactored once we can gather remote files since it is possible that we fail to gather a remote file (so we need to add it to the list of user actions) while we successfully gather other files.

Reply:
I agree that this adds a lot of complexity to `WLSDeployArchiveFile`, and will really blow up once we try the SSH stuff. Maybe an accessor class with matching `add*()` functions could handle the extra decision making and reporting, and only call `WLSDeployArchiveFile` methods when we're really adding. That would minimize the impact to existing discover code.
The accessor class methods could also take string arguments and pass java.io.File objects to the corresponding `WLSDeployArchiveFile` methods. Maybe `ArchiveUpdater`?